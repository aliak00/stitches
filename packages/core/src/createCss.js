import Object, { create } from './Object'
import createGetComputedCss from './createGetComputedCss'
import getCustomProperties from './getCustomProperties'
import getHashString from './getHashString'

/** @typedef {{ [condition: string]: string }} Conditions */
/** @typedef {{ [name: string]: number | string | Styles, when: { [condition: string]: Styles } }} Styles */
/** @typedef {{ [scale: string]: { [token: string]: number | string } }} Theme */
/** @typedef {{ [property: string]: string }} ThemeMap */
/** @typedef {{ [name: string]: (value: any) => number | string | Styles }} Utils */
/** @typedef {{ [name: string]: { [value: string]: Styles } }} Variants */

/** Returns a new styled sheet and accompanying API. */
const createCss = (
	/** @type {{ conditions?: Conditions, prefix?: string, theme?: Theme, themeMap?: ThemeMap, utils?: Utils }} Configuration */
	init,
) => {
	init = Object(init)

	/** Named conditions (media and support queries). */
	const conditions = Object(init.conditions)

	/** Prefix added before all generated classes. */
	const prefix = init.prefix == null ? '' : String(init.prefix)

	/** Returns a string of unnested CSS from an object of nestable CSS. */
	const getComputedCss = createGetComputedCss(Object(init.utils), Object(init.themeMap), conditions)

	/** Prepares global CSS and returns a function that enables the styles on the styled sheet. */
	const createThemedRule = (
		/** @type {string} CSS Selector */
		selector,
		/** @type {Theme} Object of theme scales with inner token values. */
		theme,
	) => getComputedCss(getCustomProperties(theme), [selector], [])

	/** Returns a function that enables the styles on the styled sheet. */
	const createGlobalRule = (
		/** @type {Styles} Styles representing global CSS. */
		initStyles,
	) => {
		/** @type {string[]} List of global import styles. */
		let importRuleCssTexts = []

		/** @type {string[]} List of global styles. */
		let globalRuleCssTexts = []

		for (const name in initStyles) {
			/** @type {number | string | Styles} */
			const style = initStyles[name]

			if (name === '@import') {
				importRuleCssTexts.push(getComputedCss({ [name]: style }, [], []))
			} else if (name.charCodeAt(0) === 64) {
				globalRuleCssTexts.push(getComputedCss(style, [], [name]))
			} else {
				globalRuleCssTexts.push(getComputedCss(style, [name], []))
			}
		}

		return [importRuleCssTexts, globalRuleCssTexts]
	}

	/** Prepares component CSS and returns a function that enables the styles on the styled sheet. */
	const createStyledRule = (
		/** @type {Styles & { variants?: Variants }} Styles representing component CSS. */
		initStyles,
	) => {
		// @todo implement extensions
		let { variants: variantsStyles, ...styles } = initStyles

		/** @type {string} Unique classname representing the current styled rule. */
		const styledRuleClassName = prefix + getHashString(styles)

		/** @type {string} CSS styles representing the current styled rule. */
		const styledRuleCssText = getComputedCss(styles, ['.' + styledRuleClassName], [])

		/** @type {{ [name: string]: { [value: string]: { className: string, cssText: string, when: { [key: string]: { className: string, cssText: string } } } } }} */
		const styledRuleVariants = create(null)

		/** @type {{ [key: string]: string }} */
		const inlineCssTextByClassName = create(null)

		for (const name in variantsStyles) {
			styledRuleVariants[name] = create(null)

			for (const value in variantsStyles[name]) {
				const variantStyles = variantsStyles[name][value]
				const variantClassName = styledRuleClassName + '-' + getHashString(variantStyles)

				styledRuleVariants[name][value] = [variantClassName, getComputedCss(variantStyles, ['.' + variantClassName], []), create(null)]
			}
		}

		/** Returns an expression of the current styled rule. */
		const styledRule = (
			/** @type {object} Props used to determine the expression of the current styled rule. */
			initProps,
		) => {
			const { css: inlineStyles, ...props } = Object(initProps)

			const classNames = props.className == null ? [styledRuleClassName] : String(props.className).trim().split(/\s+/).concat(styledRuleClassName)
			const variantCssTexts = []

			for (const propName in props) {
				if (propName in styledRuleVariants) {
					const styledRuleVariant = styledRuleVariants[propName]
					const propValue = props[propName]

					delete props[propName]

					if (propValue in styledRuleVariant) {
						const [className, cssText] = styledRuleVariant[propValue]

						classNames.push(className)
						variantCssTexts.push(cssText)
					} else {
						for (const innerPropName in propValue) {
							const condition = conditions[innerPropName] || innerPropName
							const innerPropValue = propValue[innerPropName]

							if (innerPropValue in styledRuleVariant) {
								const [, , conditionalVariantRules] = styledRuleVariant[innerPropValue]

								if (!(condition in conditionalVariantRules)) {
									/** @type ... */
									const conditionalVariantStyles = {
										[condition]: variantsStyles[propName][innerPropValue],
									}

									/** @type ... */
									const className = styledRuleClassName + '-' + getHashString(conditionalVariantStyles)

									conditionalVariantRules[condition] = [className, getComputedCss(conditionalVariantStyles, ['.' + className], [])]
								}

								const [className, cssText] = conditionalVariantRules[condition]

								classNames.push(className)
								variantCssTexts.push(cssText)
							}
						}
					}
				}
			}

			let inlineCssText = ''

			if (inlineStyles) {
				const inlineRuleClassName = getHashString(inlineStyles)

				classNames.push(inlineRuleClassName)

				if (!(inlineRuleClassName in inlineCssTextByClassName)) {
					inlineCssTextByClassName[inlineRuleClassName] = getComputedCss(inlineStyles, ['.' + inlineRuleClassName], [])
				}

				inlineCssText = inlineCssTextByClassName[inlineRuleClassName]
			}

			props.className = classNames.join(' ')

			return [props, styledRuleCssText, variantCssTexts, inlineCssText]
		}

		styledRule.className = styledRuleClassName

		return styledRule
	}

	return [createThemedRule, createGlobalRule, createStyledRule]
}

export default createCss
