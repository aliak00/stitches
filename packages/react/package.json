{
  "name": "@stitches/internal-react",
  "version": "0.1.0-canary.1",
  "license": "MIT",
  "source": "src/index.ts",
  "main": "dist/stitches-internal-react.cjs.js",
  "module": "dist/stitches-internal-react.esm.js",
  "typings": "dist/stitches-internal-react.cjs.d.ts",
  "files": [
    "dist",
    "src"
  ],
  "dependencies": {
    "@stitches/internal-core": "0.1.0-canary.1"
  },
  "devDependencies": {
    "react": "17.0.1"
  },
  "peerDependencies": {
    "react": "17-18"
  },
  "publishConfig": {
    "access": "restricted"
  },
  "private": true
}
