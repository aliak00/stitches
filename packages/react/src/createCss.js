import Object, { assign, create } from '../core/Object'
import createCoreCss from '../core/createCss'
import defaultThemeMap from '../core/defaultThemeMap'
import reactElementPrototype from './reactElementPrototype'

const createCss = (init) => {
	const [createThemedRule, createGlobalRule, createStyledRule] = createCoreCss(
		assign(
			{
				themeMap: defaultThemeMap,
			},
			init,
		),
	)

	let sheetSet

	let sheetParent

	let sheetTarget

	let importTextNode = new Text('')
	let globalTextNode = new Text('')
	let themedTextNode = new Text('')
	let styledTextNode = new Text('')

	const sync = () => {
		if (!sheetParent) sheetParent = document.head || document.documentElement
		if (!sheetTarget) sheetTarget = document.getElementById('stitches') || assign(document.createElement('style'), { id: 'stitches' })
		if (!sheetTarget.parentNode) sheetParent.prepend(sheetTarget)
	}

	const clear = () => {
		sync()

		sheetSet = create(null)

		sheetTarget.textContent = ''
		// sheetTarget.append(importTextNode, new Text(themedCssText), globalTextNode, themedTextNode, styledTextNode)
		sheetTarget.append(importTextNode, globalTextNode, themedTextNode, styledTextNode)

		return ''
	}

	clear.toString = clear

	clear()

	/** Returns a function that enables the styles on the styled sheet. */
	const global = (
		/** Styles representing global CSS. */
		initStyles
	) => {
		const [importRuleCssTexts, globalRuleCssTexts] = createGlobalRule(initStyles)

		let importCssText = ''
		let globalCssText = ''

		const globalRule = () => {
			for (const importRuleCssText of importRuleCssTexts) {
				if (!sheetSet[importRuleCssText]) {
					sheetSet[importRuleCssText] = true

					importCssText += importRuleCssText
				}
			}

			for (const globalRuleCssText of globalRuleCssTexts) {
				if (!sheetSet[globalRuleCssText]) {
					sheetSet[globalRuleCssText] = true

					globalCssText += globalRuleCssText
				}
			}

			if (importCssText) importTextNode.before(importCssText)

			if (globalCssText) globalTextNode.before(globalCssText)

			sync()

			return ''
		}

		globalRule.toString = globalRule

		return globalRule
	}

	const tokenPrototype = create(String.prototype, {
		computedValue: {
			get(): string {
				return 'var(' + this.variable + ')'
			},
		},
		variable: {
			get(): string {
				return '--' + this.scale + '-' + this.token
			},
		},
		toString: {
			value(): string {
				return this.computedValue
			},
		},
	})

	const theme = (
		/** CSS Class name. */
		className,
		/** Object of theme scales with inner token values. */
		themeStyles,
	) => {
		/** CSS Selector */
		const selector = className.replace(/^[A-Za-z-]/, '.$&')

		/** CSS styles representing the current theme rule. */
		const themedCssText = createThemedRule(selector, themeStyles)

		const themeRule = () => {
			if (!sheetSet[themedCssText]) {
				sheetSet[themedCssText] = true

				themedTextNode.before(themedCssText)

				sync()
			}

			return themeRule
		}

		themeRule.toString = () => selector

		for (const scale in themeStyles) {
			themeRule[scale] = create(null)

			for (const token in themeStyles[scale]) {
				themeRule[scale][token] = create(tokenPrototype, {
					value: { value: themeStyles[scale][token] },
					scale: { value: scale },
					token: { value: token },
				})
			}
		}

		return themeRule
	}

	assign(theme, theme(':root', init.theme)())

	const styled = new Proxy(
		(
			/** Type of component element. */
			type = 'span',
			/** Styles representing component CSS. */
			styles,
		) => {
			const styledRule = createStyledRule(Object(styles))
			const source = create(reactElementPrototype, { type: { value: type } })
			const selector = '.' + styledRule.className

			let styledRuleNode = new Text('')
			let varianRuleNode = new Text('')
			let inlineRuleNode = new Text('')

			styledTextNode.before(styledRuleNode, varianRuleNode, inlineRuleNode)

			const styledComponent = (init) => {
				const [props, styledRuleCssText, varianCssTexts, inlineCssText] = styledRule(init)

				if (!sheetSet[styledRuleCssText]) {
					sheetSet[styledRuleCssText] = true

					styledRuleNode.before(styledRuleCssText)
				}

				for (const varianCssText of varianCssTexts) {
					if (!sheetSet[varianCssText]) {
						sheetSet[varianCssText] = true

						varianRuleNode.before(varianCssText)
					}
				}

				if (!sheetSet[inlineCssText]) {
					sheetSet[inlineCssText] = true

					inlineRuleNode.before(inlineCssText)
				}

				sync()

				const element = create(source, { props: { value: props } })

				element.toString = () => selector

				return element
			}

			styledComponent.toString = () => String(styledComponent())

			return Object.setPrototypeOf(styledComponent, Object(type))
		},
		{
			get: (target, type) => (type in target ? (typeof target[type] === 'function' ? target[type].bind(target) : target[type]) : target.bind(null, type)),
		},
	)

	return { clear, global, styled, theme }
}

export default createCss
