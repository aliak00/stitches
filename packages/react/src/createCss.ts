import Object, { assign, create } from '@stitches/core/Object'
import createCoreCss from '@stitches/core/createCss'
import defaultThemeMap from '@stitches/core/defaultThemeMap'
import reactElementPrototype from './reactElementPrototype'
import ThemeToken from './ThemeToken'

const createCss = (init: any) => {
	const [createThemedRule, createGlobalRule, createStyledRule] = createCoreCss(
		assign(
			{
				themeMap: defaultThemeMap,
			},
			init,
		),
	)

	let sheetSet: { [key: string]: true }

	let sheetParent: HTMLHeadElement | HTMLHtmlElement

	let sheetTarget: HTMLStyleElement

	let importTextNode = new Text('')
	let globalTextNode = new Text('')
	let themedTextNode = new Text('')
	let styledTextNode = new Text('')

	const sync = () => {
		if (!sheetParent) sheetParent = document.head || document.documentElement
		if (!sheetTarget) sheetTarget = <HTMLStyleElement>document.getElementById('stitches') || assign(document.createElement('style'), { id: 'stitches' })
		if (!sheetTarget.parentNode) sheetParent.prepend(sheetTarget)
	}

	const clear = () => {
		sync()

		sheetSet = create(null)

		sheetTarget.textContent = ''
		// sheetTarget.append(importTextNode, new Text(themedCssText), globalTextNode, themedTextNode, styledTextNode)
		sheetTarget.append(importTextNode, globalTextNode, themedTextNode, styledTextNode)

		return ''
	}

	clear.toString = clear

	clear()

	/** Returns a function that enables the styles on the styled sheet. */
	const global = (
		/** Styles representing global CSS. */
		initStyles: Styles,
	) => {
		const [importRuleCssTexts, globalRuleCssTexts] = createGlobalRule(initStyles)

		let importCssText = ''
		let globalCssText = ''

		const globalRule = () => {
			for (const importRuleCssText of importRuleCssTexts) {
				if (!sheetSet[importRuleCssText]) {
					sheetSet[importRuleCssText] = true

					importCssText += importRuleCssText
				}
			}

			for (const globalRuleCssText of globalRuleCssTexts) {
				if (!sheetSet[globalRuleCssText]) {
					sheetSet[globalRuleCssText] = true

					globalCssText += globalRuleCssText
				}
			}

			if (importCssText) importTextNode.before(importCssText)

			if (globalCssText) globalTextNode.before(globalCssText)

			sync()

			return ''
		}

		globalRule.toString = globalRule

		return globalRule
	}

	/** Returns a function that assigns a themed rule to the styled sheet. */
	const theme = (
		/** CSS Class name. */
		className: string,
		/** Object of theme scales with inner token values. */
		themeStyles: Theme,
	) => {
		/** CSS Selector */
		const selector = className.replace(/^[A-Za-z-]/, '.$&')

		/** CSS styles representing the current theme rule. */
		const themedCssText = createThemedRule(selector, themeStyles)

		const themeRule = () => {
			if (!sheetSet[themedCssText]) {
				sheetSet[themedCssText] = true

				themedTextNode.before(themedCssText)

				sync()
			}

			return themeRule
		}

		themeRule.toString = () => selector

		for (const scale in themeStyles) {
			themeRule[scale] = create(null)

			for (const token in themeStyles[scale]) {
				themeRule[scale][token] = assign(new ThemeToken(), {
					value: themeStyles[scale][token],
					scale: scale,
					token: token,
				})
			}
		}

		return themeRule
	}

	assign(theme, theme(':root', init.theme)())

	/** Returns a function that assigns a keyframe rule to the styled sheet. */
	const keyframes = (
		/** Keyframe name. */
		name: string,
		/** Styles representing global CSS. */
		initStyles: Styles,
	) => {
		name = String(name)

		const [, [keyframeRuleCssText]] = createGlobalRule({
			['@keyframes ' + name]: initStyles,
		})

		const keyframesRules = () => {
			globalTextNode.before(keyframeRuleCssText)

			sync()

			return name
		}

		keyframesRules.toString = keyframesRules

		return keyframesRules
	}

	const styled = new Proxy(
		(
			/** Type of component element. */
			type = 'span',
			/** Styles representing component CSS. */
			styles: Styles & { variants: Variants },
		) => {
			const styledRule = createStyledRule(Object(styles))
			const source = create(reactElementPrototype, { type: { value: type } })
			const selector = '.' + styledRule.className

			let styledRuleNode = new Text('')
			let varianRuleNode = new Text('')
			let inlineRuleNode = new Text('')

			styledTextNode.before(styledRuleNode, varianRuleNode, inlineRuleNode)

			const styledComponent = (init: any) => {
				const [props, styledRuleCssText, varianCssTexts, inlineCssText] = styledRule(init)

				if (!sheetSet[styledRuleCssText]) {
					sheetSet[styledRuleCssText] = true

					styledRuleNode.before(styledRuleCssText)
				}

				for (const varianCssText of varianCssTexts) {
					if (!sheetSet[varianCssText]) {
						sheetSet[varianCssText] = true

						varianRuleNode.before(varianCssText)
					}
				}

				if (!sheetSet[inlineCssText]) {
					sheetSet[inlineCssText] = true

					inlineRuleNode.before(inlineCssText)
				}

				sync()

				const element = create(source, { props: { value: props } })

				element.toString = () => selector

				return element
			}

			styledComponent.toString = () => String(styledComponent())

			return Object.setPrototypeOf(styledComponent, Object(type))
		},
		{
			get: (target, type) => (type in target ? (typeof target[type] === 'function' ? target[type].bind(target) : target[type]) : target.bind(null, type)),
		},
	)

	return { clear, global, keyframes, styled, theme }
}

export default createCss
