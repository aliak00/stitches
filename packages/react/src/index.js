import Object, { assign } from '../../core/src/Object.js'
import createCoreCss from '../../core/src/index.js'
import defaultThemeMap from '../../core/src/defaultThemeMap.js'

const $$typeof = Symbol.for('react.element')

export default (init) => {
	const hasDocument = typeof document === 'object'

	const importText = hasDocument && new Text('')
	const globalText = hasDocument && new Text('')
	const themedText = hasDocument && new Text('')
	const styledText = hasDocument && new Text('')

	const createOnChange = hasDocument ? textNode => data => (textNode.data = data) : () => undefined

	let sheetParent
	let sheetTarget

	init = assign(
		{
			onImport: createOnChange(importText, 'import'),
			onGlobal: createOnChange(globalText, 'global'),
			onThemed: createOnChange(themedText, 'themed'),
			onStyled: createOnChange(styledText, 'styled'),
			onResets() {
				if (hasDocument) {
					this.sync()

					sheetTarget.textContent = importText.data = globalText.data = themedText.data = styledText.data = ''
					sheetTarget.append(importText, globalText, themedText, styledText)
				}
			}
		},
		init,
	)

	const sheet = createCoreCss(init)

	return assign(sheet, {
		sync() {
			if (hasDocument) {
				if (!sheetParent) sheetParent = document.head || document.documentElement
				if (!sheetTarget) sheetTarget = document.getElementById('stitches') || assign(document.createElement('style'), { id: 'stitches' })
				if (!sheetTarget.parentNode) sheetParent.prepend(sheetTarget)
			}
		},
		styled: new Proxy(
			/** Returns a React component. */
			(
				/** Type of component element. */
				asType = 'span',
				/** Styles representing component CSS. */
				initStyles,
			) => {
				const expression = sheet.css(asType, initStyles)

				/** Returns a React element. */
				return Object.setPrototypeOf(
					assign((
						/** Props used to determine the expression of the current styled rule. */
						initProps,
					) => {
						// extract props, individually extracting `as` & `ref` props
						const {
							props: { as: type = asType, ref = null, ...props },
							...expressedProps
						} = expression(initProps)

						// sync the dynamic stylesheet
						sheet.sync()

						// return the react element
						return { ...expressedProps, constructor: undefined, $$typeof, props, ref, type, __v: 0 }
					}, expression),
					// extends the asType, otherwise Object
					Object(asType)
				)
			},
			{
				get: (target, type) => (type in target ? (typeof target[type] === 'function' ? target[type].bind(target) : target[type]) : target.bind(null, type)),
			},
		),
	}).reset()
}

export { defaultThemeMap }
