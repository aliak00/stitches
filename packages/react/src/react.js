import createCore, { _StyledSheet } from '@stitches/core'

/**
 * Returns a new StyledSheet.
 * @type {import('./types').ReactFactory}
 */
const reactFactory = (
	/** Configuration of the Stitches instance. */
	init,
) => {
	let sheetHeadElement
	let sheetStyleElement
	let sheetGlobalTextNode
	let sheetStyledTextNode
	let sheetThemedTextNode

	const hasDocument = typeof document === 'object'

	if (hasDocument) {
		sheetGlobalTextNode = new Text('')
		sheetStyledTextNode = new Text('')
		sheetThemedTextNode = new Text('')
		sheetStyleElement = document.createElement('style')
		sheetHeadElement = document.head

		sheetStyleElement.append(sheetGlobalTextNode)
		sheetStyleElement.append(sheetThemedTextNode)
		sheetStyleElement.append(sheetStyledTextNode)
		sheetHeadElement.append(sheetStyleElement)
	}

	const core = hasDocument
		? createCore({
				...Object(init),
				onGlobal(cssOfImportRules, cssOfGlobalRules) {
					sheetGlobalTextNode.data = cssOfImportRules + cssOfGlobalRules
				},
				onStyled(cssOfStyledRules) {
					sheetStyledTextNode.data = cssOfStyledRules
				},
				onThemed(cssOfTheme, cssOfThemedRules) {
					sheetThemedTextNode.data = cssOfTheme + cssOfThemedRules
				},
		  })
		: createCore(init)

	const $$typeof = Symbol.for('react.element')
	const $$styled = Symbol.for('stitches.component')

	core.styled = (type, init) => {
		const component = Object.assign(props => render(props), {
			[$$styled]: true,
			displayName: 'StitchesComponent',
			rule: type[$$styled]
				? core.css(
						{
							classNames: type.rule.classNames,
							variants: type.rule.variants,
						},
						init,
				  )
				: core.css(init),
			type: type[$$styled] ? type.type : type,
		})

		const render = ({ as, className: classNameOverride, ref = null, ...init }) => {
			const classNameOverrides = String(classNameOverride || '').split(/\s+/)
			const expression = component.rule(init)

			const { classNames, props } = expression

			return {
				$$typeof,
				key: null,
				props: {
					...props,
					className: classNames.concat(classNameOverrides).filter(Boolean).join(' '),
				},
				ref,
				type: as || component.type,
				_owner: null,
			}
		}

		return component
	}

	return core
}

export default reactFactory
