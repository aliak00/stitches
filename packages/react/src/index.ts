import createCss from '@stitches/core'
import define from './lib/define'

/** Factory that returns a StyledSheet. */
function createStyled(init?: StyledSheetFactoryInit & { onUpdate(sheet: StyledSheet): void }) {
	const sheet = createCss(init)
	const { onUpdate } = Object(init)

	return define(sheet, {
		styled(type: string | object, init: any & object) {
			const createExpression = sheet.css(init)

			return function Component({ as, className: classNameOverride, ref = null, ...props }: object & any) {
				const classNameOverrides = String(classNameOverride || '')
					.split(/\s+/)
					.filter(Boolean)

				const expression = createExpression(props)

				props.className = [...expression.classNames, ...classNameOverrides].join(' ')

				if (typeof onUpdate === 'function') {
					onUpdate(sheet)
				}

				return {
					$$typeof: Symbol.for('react.element'),
					key: null,
					props,
					ref,
					type: as || type,
					_owner: null,
				}
			}
		},
	})
}

export default createStyled
