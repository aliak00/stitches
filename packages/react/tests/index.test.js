import * as react from 'react'
import * as renderer from 'react-test-renderer'
import createCss from '../src/index.js'

describe('Stitches React', () => {
	let wrapper = renderer.create(react.createElement(react.Fragment))
	const sheet = createCss()
	let styled = sheet.styled
	let Button = styled('button', {
		backgroundColor: 'gainsboro',
		borderRadius: '9999px',
		fontWeight: 500,
		padding: '0.75em 1em',
		border: 0,
		transition: 'all 200ms ease',

		'&:hover': {
			transform: 'translateY(-2px)',
			boxShadow: '0 10px 25px rgba(0, 0, 0, .3)',
		},
	})

	// test('createCss returns an object with a "styled" function', () => {
	// 	expect(sheet.styled).toBeInstanceOf(Function)
	// })

	// test('"styled" function returns a Button component', () => {
	// 	expect(Button).toBeInstanceOf(Function)
	// })
})
