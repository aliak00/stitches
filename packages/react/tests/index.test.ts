// import { create, act } from 'react-test-renderer'
import createCss from '../src/index'

// void create
// void act
// void createCss

describe('Stitches React', () => {
	const { styled } = createCss({
		theme: {
			color: {
				lite: 'gainsboro',
			},
			radii: {
				full: '9999px',
			},
			space: {
				full: '1em',
				tenSixteenth: 'calc(10em / 16)',
			},
		},
	})

	test('basic functionality', () => {
		expect(styled).toBeInstanceOf(Function)

		const Button = styled('button', {
			backgroundColor: '$lite',
			borderRadius: '$full',
			fontWeight: 500,
			padding: '$tenSixteenth $full',
			border: 0,
			transition: 'all 200ms ease',

			'&:hover': {
				transform: 'translateY(-2px)',
				boxShadow: '0 10px 25px rgba(0, 0, 0, .3)',
			},
		})

		expect(Button).toBeInstanceOf(Function)

		const Expression = Button({})

		expect(Expression).toStrictEqual({
			$$typeof: Symbol.for('react.element'),
			key: null,
			props: { className: 's-2j80d3' },
			ref: null,
			type: 'button',
			_owner: null,
		})
	})
})

// 	// create button, write css
// 	const expectHML1 = `<button class="s-2j80d3">Styled Button</button>`
// 	const resultHML1 = `<button class="${buttonClass}">Styled Button</button>`

// 	expect(resultHML1).toBe(expectHML1)

// 	const darkButtonClass = css(buttonClass, {
// 		backgroundColor: 'black',
// 		color: 'white',
// 	})

// 	// create button, write css
// 	const expectHML2 = `<button class="s-ohkga5 s-2j80d3">Styled Button</button>`
// 	const resultHML2 = `<button class="${darkButtonClass}">Styled Button</button>`

// 	expect(resultHML2).toBe(expectHML2)
// })
