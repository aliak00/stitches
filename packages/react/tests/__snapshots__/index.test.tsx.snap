// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled Breakpoints work in variants 1`] = `
{
  "orderedAppliedStyles": [
    "._initial_c_vfarC {color: red;}",
    "._breakpointOne_h_KwWJf {height: 10px;}",
    "._breakpointTwo_h_hiyRrM {height: 20px;}"
  ],
  "props": {
    "className": "_initial_c_vfarC _breakpointOne_h_KwWJf _breakpointTwo_h_hiyRrM scid-kvQUME"
  },
  "children": [
    "no variant"
  ]
}
`;

exports[`styled Breakpoints work in variants and when a responsive variant is passed ot the element 1`] = `
{
  "orderedAppliedStyles": [
    "._initial_c_vfarC {color: red;}",
    "._breakpointOne_h_dDcWdv {height: 100px;}",
    "._breakpointTwo_h_dPiPAc {height: 200px;}"
  ],
  "props": {
    "className": "_breakpointOne_h_dDcWdv _breakpointTwo_h_dPiPAc _initial_c_vfarC scid-cXpAaU"
  },
  "children": [
    "with responsive variant"
  ]
}
`;

exports[`styled Breakpoints work in variants and when the variant is passed to the jsx element 1`] = `
{
  "orderedAppliedStyles": [
    "._initial_c_vfarC {color: red;}",
    "._breakpointOne_h_kgNeVv {height: 1000px;}",
    "._breakpointTwo_h_fptqfg {height: 2000px;}"
  ],
  "props": {
    "className": "_breakpointOne_h_kgNeVv _breakpointTwo_h_fptqfg _initial_c_vfarC scid-cXpAaU"
  },
  "children": [
    "with variant"
  ]
}
`;

exports[`styled Handles css prop 1`] = `
{
  "orderedAppliedStyles": [
    "._fNPZtH {background-color: gray;}",
    "._vfarC {color: red;}"
  ],
  "props": {
    "className": "_vfarC _fNPZtH scid-dmZJpl"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled Handles css prop 2`] = `
{
  "orderedAppliedStyles": [
    "._fNPZtH {background-color: gray;}",
    "._vfarC {color: red;}"
  ],
  "props": {
    "className": "_vfarC _fNPZtH scid-dmZJpl"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled It handles variants with a numeric value of 0 1`] = `
{
  "orderedAppliedStyles": [
    "._NkcBC {height: 1px;}"
  ],
  "props": {
    "className": "_NkcBC scid-fUwIQk"
  },
  "children": [
    "height should equal '1px'"
  ]
}
`;

exports[`styled Renders basic component 1`] = `
{
  "orderedAppliedStyles": [
    "._dawVaV div {font-size: 18px;}"
  ],
  "props": {
    "className": "hi _dawVaV scid-dIyoRL"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled Renders components from other libraries 1`] = `
{
  "orderedAppliedStyles": [
    "._vfarC {color: red;}"
  ],
  "props": {
    "className": "_vfarC scid-gmqXFB"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled Renders composed component 1`] = `
{
  "orderedAppliedStyles": [
    "._genYDa {color: black;}",
    "._bQrNcz {font-family: monospace;}"
  ],
  "props": {
    "className": "scid-kngyIZ _bQrNcz _genYDa scid-dvWlMZ"
  },
  "children": [
    "console.log"
  ]
}
`;

exports[`styled renders component with variant 1`] = `
{
  "orderedAppliedStyles": [
    "._fNPZtH {background-color: gray;}"
  ],
  "props": {
    "className": "_fNPZtH scid-gFxMiu"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled renders component with variant 2`] = `
{
  "orderedAppliedStyles": [
    "._gwFOTp {background-color: red;}"
  ],
  "props": {
    "className": "_gwFOTp scid-gFxMiu"
  },
  "children": [
    "hello world"
  ]
}
`;

exports[`styled renders component with variant 3`] = `
{
  "orderedAppliedStyles": [
    "._htDnsg {background-color: blue;}"
  ],
  "props": {
    "className": "_htDnsg scid-gFxMiu"
  },
  "children": [
    "hello world"
  ]
}
`;
